### DISCUSSION TITLE
Auto determine Client

### DISCUSSION THREAD

**Discussion by EVB1980** [belkiss/niftyplugins#18] [Ideas]
Hi, sorry for disturb, I have tryed your plugin, changed it in some and I think I have good improvement for plugin, which could be useful for who is using multiple workspaces.

DETAILS:
In options of plugin we should setup 'Client' property, but if in same time we wroking in different Clients from several instances of VS we cant have different settings (the last setting is stored).

I've made some code modification wich determines Client setting by file name of file wich should be changed, please look attached file [niftyplugins.zip](https://github.com/belkiss/niftyplugins/files/7602070/niftyplugins.zip), and if you will found it useful - then please integrate to your main branch.

**Comment by belkiss** [comment]
Heya!
Thanks for trying out the plugin and opening a discussion, you're definitely not disturbing :)
What I usually do is to create p4config.txt files (and have the environment variable `P4CONFIG=p4config.txt`) ([doc](https://www.perforce.com/manuals/cmdref/Content/CmdRef/P4CONFIG.html)).
I put one of those files at the root of every workspace I have on my machine, so I don't need to set anything in nifty settings, it just works :) Let me know if that works for you too.

I'll have a look at your changes soon and let you know.

**Reply by belkiss** [reply]
Side note, could you send those as a MR on github?

**Reply by EVB1980** [reply]
This will not work if you will work with streams, if you didn't work with that feature of perforce - it is just like branches of same code in differemt folders, for example we have folders structure:

D:\P4\
      ws_A\....
      ws_B\....
      
ws_A and ws_B - are physical folders of workspaces (or clients), each workspace looks to own stream, and one of streams is parent and another is child, for simplicity let they have same names as workspaces directories, and ws_A is parent stream of ws_B

In this case environment variables will not work because p4config.txt in both directories will be same (with same text in it, because branch inheritance).

Streams graph example:
![image](https://user-images.githubusercontent.com/8556910/143428303-f63db86b-36ea-4367-b6bd-2e0b521ea4b6.png)

**Reply by belkiss** [reply]
Hum, I do use streams (albeit rarely swap my workspace), but if I understand correctly you have sthg like:

```
D:\P4\ws_A\some_file.cpp <- belongs to ws_A
D:\P4\ws_B\some_file.cpp <- belongs to ws_B
```

So if you do:
```
D:\P4\ws_A\p4config.txt
D:\P4\ws_B\p4config.txt
```

The ws_A p4config.txt content would look something like:
```
P4PORT=the-server-address:4777
P4USER=user.name
P4CLIENT=ws_A
P4IGNORE=.gitignore;p4ignore.txt
```
(same for ws_B except the P4CLIENT line).

The only env var you have is P4CONFIG, telling p4 to look up the name that is set as a value for p4config files.

**Reply by EVB1980** [reply]
Yes, all as you say, but this will work only if you will not comit p4config.txt to perforce, but if so we can have folowing problem:

If we working in team, each member should create p4config.txt for each directory of client (on different machines perforce mappings are different), and if you have many clients (or streams) - it is difficult, all programmers are lazy persons in good meaning of this word))).

And really the code I've made will free you from creating such files, you should specify only settings:
![image](https://user-images.githubusercontent.com/8556910/143430392-0bd362ba-3698-4321-b113-623be02b51f5.png)

**Reply by EVB1980** [reply]
I'm not sure in one thing, when I retrieving workspaces information from perforce server, it would be better to filter by machine, to not get mine workspaces from another machines (because they could have same root directories, but it is really rare case).

**Reply by EVB1980** [reply]
And I'm sorry, but I realy do not understand what is MainLinePath, is it for merging from child to parent?
I working with perforce about 7 years (4 years with streams), but maybe I missing ssomething.

**Reply by belkiss** [reply]
I don't think it's such a pain to create such local files, since you only need to do it once when you create a new workspace, which is (at least for me!) not that often, and also it has the additional benefit to have all the other p4 commands work from cmd line, not only nifty!
And yes, of course they can't be submitted since they are user specific.

> would be better to filter by machine, to not get mine workspaces from another machines 

Agreed, but note that the host field is optional, so you would need to take this into account.

I'm thinking that I'd be open to having nifty generate the p4config.txt files, that way your workspace discovery could run on demand and not each time?

> what is MainLinePath

That's a legacy feature from the original nifty (that I don't use myself :') ), but the idea is to set the depot path to the branch you consider the main one, so if your workspace is set to a different one you can make p4 queries for the corresponding file on main, for instance diff your branch version against main, etc.

**Reply by EVB1980** [reply]
Sorry fr long reply...

"...that way your workspace discovery could run on demand and not each time?"

It could be done on solution open, one time.

But I didn't work with VS extensions, so I do not know how to do this in code. :(

**Comment by belkiss** [comment]
No worries! Yes that would work, if we find a P4CONFIG file but no file that it indicates, we could auto create it. Not sure about when though, I'm not a fan of adding things when the solution is loaded (it's already slow enough 😅), but we could do that check on the first perforce operation the user does?
